File: haskell/bob/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import Data.Char
import qualified Data.Text as T
import Data.Text (Text)

isQuestion :: Text -> Bool
isQuestion = T.isSuffixOf "?" . T.strip

isLoud :: Text -> Bool
isLoud sentence =
  let s = (T.filter isAlpha sentence)
  in T.all (\c -> isUpper c) s && not (T.null s)

isSilence :: Text -> Bool
isSilence = T.all isSpace

responseFor :: Text -> Text
responseFor sentence
  | isQuestion sentence && isLoud sentence = "Calm down, I know what I'm doing!"
  | isQuestion sentence = "Sure."
  | isLoud sentence = "Whoa, chill out!"
  | isSilence sentence = "Fine. Be that way!"
  | otherwise = "Whatever."

File: zig/bob/bob.zig
const std = @import("std");

pub fn response(s: []const u8) []const u8 {
    const question = isQuestion(s);

    if (isSilent(s)) {
        return "Fine. Be that way!";
    } else if (isShouting(s)) {
        if (question) {
            return "Calm down, I know what I'm doing!";
        } else {
            return "Whoa, chill out!";
        }
    } else if (question) {
        return "Sure.";
    }

    return "Whatever.";
}

fn isQuestion(s: []const u8) bool {
    var question = false;
    for (s) |char| {
        if (std.ascii.isAlpha(char)) {
            question = false;
        } else if (char == '?') {
            question = true;
        }
    }

    return question;
}

fn isShouting(s: []const u8) bool {
    var shouting = false;
    for (s) |char| {
        if (std.ascii.isUpper(char)) {
            shouting = true;
        } else if (std.ascii.isLower(char)) {
            return false;
        }
    }

    return shouting;
}

fn isSilent(array: []const u8) bool {
    for (array) |element| {
        if (!std.ascii.isWhitespace(element)) {
            return false;
        }
    }

    return true;
}
